fastar/rules.py

```diff
@@ -273,34 +252,40 @@ def dot_general_dependency_rule(
               materialize(rhs_count)
               * prod(np.take(outshape, lhs_other_out_dims))]
   return ([lhs_box, rhs_box], incounts,
-          lambda *inslices: lax.dot_general(
-              *inslices, dimension_numbers, precision))
+          (dict(dimension_numbers=dimension_numbers, precision=precision), None))
 
 dependency_rules[lax.dot_general_p] = dot_general_dependency_rule
+kernels[lax.dot_general_p] = lax.dot_general
+
+def ceil_divide(x1, x2):
+  return -np.floor_divide(np.negative(x1), x2)
 
 def pad_dependency_rule(outstart, outcount, operand, padding_value, padding_config):
   lo, _, interior = unzip3(padding_config)
   dilation = np.array(interior) + 1
   outstart_lo = np.subtract(outstart, lo)
   inclip = lambda indices: np.clip(indices, 0, operand.shape)
-  instart = inclip(lax.lax._ceil_divide(outstart_lo, dilation))
-  instop = inclip(lax.lax._ceil_divide(outstart_lo + outcount.shape, dilation))
+  instart = inclip(ceil_divide(outstart_lo, dilation))
+  instop = inclip(ceil_divide(outstart_lo + outcount.shape, dilation))
   inshape = instop - instart
   insize = prod(inshape)
   offset = instart * dilation - outstart_lo
   limit = offset + np.maximum(0, (np.array(inshape) - 1) * dilation + 1)
   incount = Ones(inshape) if is_ones(outcount) else laxref.slice(
     outcount, offset, limit, dilation) if insize else None
   padcount = outcount.size - insize
-  def outslice(inslice, padding_value):
-    assert inslice is None or np.array_equal(inslice.shape, inshape)
-    return (lax.pad(inslice, padding_value,
-                    zip(offset, np.array(outcount.shape) - limit, interior))
-            if insize else jnp.full(outcount.shape, padding_value, operand.dtype))
   return ([(instart, inshape) if insize else None, ([], [])],
-          [incount, padcount], outslice)
+          [incount, padcount], (dict(
+    padding_config=tuple(zip(tuple(offset), tuple(np.array(outcount.shape) - limit), tuple(interior))),
```

Comment:
I think the `tuple`s inside the zip won't do anything (possible I'm mistaken here so apologies if so), but I think
```python
zip(tuple(offset), tuple(np.array(outcount.shape) - limit), tuple(interior))
```
would be identical to
```python
zip(offset, np.array(outcount.shape) - limit, interior)
```